src/components/infinite-scroll/InfiniteScroll.tsx

```diff
@@ -0,0 +1,108 @@
+/**
+ * From https://github.com/jaredpalmer/react-simple-infinite-scroll
+ * Updated to accept a scroll container React ref as the "window"
+ */
+import * as React from 'react';
+import throttleFn from 'lodash/throttle';
+
+export type Props = {
+    /** Components to render, should include entities/loading states/end message */
+    children: React.ReactNode;
+    /** Does the resource have more entities */
+    hasMore: boolean;
+    /** Are more entities being loaded */
+    isLoading: boolean;
+    /** Callback to load more entities */
+    onLoadMore: () => void;
+    /** React ref of the scroll container.
+     * Used to listen to scroll events and calculate onLoadMore calls.
+     * Set useWindow to true to use the window as scroll container.
+     */
+    scrollContainerNode?: HTMLElement;
+    /** Scroll threshold */
+    threshold?: number;
+    /** Throttle rate */
+    throttle?: number;
+    /** Set useWindow to true to use the window as scroll container. If set to true, will ignore scrollContainerRef. */
+    useWindow?: boolean;
+};
+
+function InfiniteScroll({
+    children,
+    hasMore,
+    isLoading,
+    onLoadMore,
+    scrollContainerNode,
+    threshold = 100,
+    throttle = 64,
+    useWindow = false,
+}: Props) {
+    const sentinelRef = React.createRef<HTMLDivElement>();
+
+    function getContainer(): EventTarget | undefined {
+        return useWindow ? window : scrollContainerNode;
+    }
+
+    React.useEffect(() => {
+        function onContainerScroll() {
+            if (isLoading || !hasMore) return;
+
+            if (sentinelRef.current == null) return;
+            const { top: sentinelTop } = sentinelRef.current.getBoundingClientRect();
+
+            if (useWindow) {
+                if (sentinelTop - window.innerHeight < threshold) {
+                    onLoadMore();
+                }
+            } else {
+                if (!scrollContainerNode) return;
+
+                const { bottom: containerBottom } = scrollContainerNode.getBoundingClientRect();
+                if (sentinelTop - containerBottom < threshold) {
+                    onLoadMore();
+                }
+            }
+        }
+        const scrollHandler = throttleFn(onContainerScroll, throttle);
+        const resizeHandler = throttleFn(onContainerScroll, throttle);
+
+        function addEventListeners() {
+            const container = getContainer();
+            if (container) {
+                container.addEventListener('scroll', scrollHandler);
+                container.addEventListener('resize', resizeHandler);
+            }
+        }
+
+        addEventListeners();
+
+        return function removeEventListeners() {
```

Comment:
This might be preference but is the named function necessary?

Reply:
you mean if the name is necessary? it's for readability, I can make it anonymous if you prefer

Reply:
For my two cents, I have never done this, but when I saw it, I thought, hey, that's not a bad idea. In this case it's pretty clear that the function removes the event listeners, but this might be a good, explicit pattern to establish.